### PR TITLE
Add calculate difficulty function

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description
Added a function to calculate the required difficulty given the time between two blocks
Refactored function name of ProofOfWork struct to more accurately reflect the work of the function and differentiate in regards to above.

## Motivation and Context
We need to be able to adjust the required difficulty as the hash rate climbs. 

## How Has This Been Tested?
Added unit tests

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/proof_of_work/difficulty.rs [hunk 1]
```diff
@@ -38,6 +38,19 @@ impl Difficulty {
     pub fn min() -> Difficulty {
         Difficulty(MIN_DIFFICULTY)
     }
+
+    /// This function will calculate the required difficulty given the time taken to calculate 2016 blocks
+    /// The time is the difference in seconds between the two headers
+    pub fn calculate_req_difficulty(time: i64, difficulty: Difficulty) -> Difficulty {
```
Comment (CjS77) [inline base_layer/core/src/proof_of_work/difficulty.rs]: suggested name: calculate_difficulty_adjustment

File: base_layer/core/src/proof_of_work/difficulty.rs [hunk 2]
```diff
@@ -38,6 +38,19 @@ impl Difficulty {
     pub fn min() -> Difficulty {
         Difficulty(MIN_DIFFICULTY)
     }
+
+    /// This function will calculate the required difficulty given the time taken to calculate 2016 blocks
+    /// The time is the difference in seconds between the two headers
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply base_layer/core/src/proof_of_work/difficulty.rs]: Yes the idea is of using the bitcoin one is just a stop gap measure until we have time to discuss a proper one.

File: base_layer/core/src/proof_of_work/difficulty.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (SWvheerden) [reply base_layer/core/src/proof_of_work/difficulty.rs]: I purposefully made it i64 as that is the unit returned by the chrono time interval. 
And if its u64, we will never by able to adjust the difficulty down, only up.

Comment (SWvheerden) [reply base_layer/core/src/proof_of_work/difficulty.rs]: Using : calculate_difficulty_adjustment suggests  to me that this function will return the amount the difficulty should be adjusted by.  The function returns the already adjusted difficulty.

Comment (CjS77) [reply base_layer/core/src/proof_of_work/difficulty.rs]: Fair enough.

[the remark this answers is not in the document]
Comment (SWvheerden) [reply base_layer/core/src/proof_of_work/difficulty.rs]: fair enough, will update the name to interval. 
I want to try and stick to i64 to improve the calling code as you would in all likelihood call DateTime.Timestamp which returns an i64

**Pull request merged**